src/core/types.rs

```diff
@@ -0,0 +1,303 @@
+use std::collections::{HashSet, HashMap, BTreeSet};
+use std::cell::{RefCell, Ref};
+use std::ops::Deref;
+use log::{debug};
+
+use pyo3::prelude::*;
+
+use crate::importlib;
+
+#[pyclass]
+#[derive(Clone)]
+pub struct DepNode {
+    #[pyo3(get)]
+    pub name: String,
+    #[pyo3(get)]
+    pub spec: importlib::ModuleSpec,
+    // The dependencies & dependents by spec.name
+    dependencies: HashSet<String>,
+    #[pyo3(get)]
+    dependents: HashSet<String>,
+    #[pyo3(get)]
+    depth: Option<i32>
+}
+
+impl DepNode {
+    pub fn new(spec: importlib::ModuleSpec, depth: Option<i32>) -> DepNode {
+        DepNode {
+            name: spec.name.clone(),
+            spec: spec,
+            dependencies: HashSet::new(),
+            dependents: HashSet::new(),
+            depth: depth // Allow for uninitialized depths
+        }
+    }
+
+    fn merge(&mut self, other: DepNode) {
+        // Few sanity checks
+        assert!(self.name == other.name);
+        if let Some(origin) = &self.spec.origin {
+            assert!(origin == &other.spec.origin.unwrap());
```

Comment:
`assert_eq!()` will print out what the two things being compared are, which can help speed up debugging

I'd also recommend doing `assert_eq!(origin, Some(other.spec.origin))` to avoid unwrapping